### ISSUE TITLE
rust engine consume a lot of memory compared to pyarrow

### ISSUE THREAD

**Issue by djouallah** [delta-io/delta-rs#2968]
# Environment

**Delta-rs version**:
0.21.0

**Binding**: 

**Environment**:
- **OS**:
Linux

***
# Bug

switching from pyarrow engine to rust increase memory usage by nearly 3X, the job used to works fine, but now, getting OOM errors.

I added a reproducible example with only 60 input files to demo the issue

https://colab.research.google.com/drive/1fahlV0FgKSAS8sQvRMu47s3bDP1ekLbb#scrollTo=333a177b-f075-412e-8ca1-32d44f8c07eb

**Comment by rtyler** [comment]
@djouallah :wave: in the attached notebook, which write_deltalake call is resulting in memory pressure?  I'm less familiar with duckdb, but I assume the `df`  objects that it is producing are pyarrow.DataSet? or are they another type?

**Comment by djouallah** [comment]
it is an Arrow RecordBatchReader I think

**Comment by ion-elgreco** [comment]
Rust engine materalizes everything to memory prior to starting the whole process. Pyarrow probably writes batch by batch

**Comment by djouallah** [comment]
just for my own understanding, is this something that can be fixed by datafusion ?

**Comment by ion-elgreco** [comment]
@djouallah there is a PR to address this but the contributor didn't have time to finish it yet: https://github.com/delta-io/delta-rs/pull/2289

**Comment by djouallah** [comment]
using 0.22, I just use the pyarrow engine, yes, processing 2300 csv files and it works great !!!
<img width="388" alt="image" src="https://github.com/user-attachments/assets/b7ccfabf-035a-4778-b607-ea84ac137ac8">

**Comment by ldacey** [comment]
Yeah, I saw the "pyarrow" deprecation message and swapped to the rust engine for one of my tables. It is only about 2.6 million rows per partition but it is very wide and all text data (200+ columns). The pods were failing due to OOM errors after doing this so I need to use the pyarrow engine and overwrite the partition still.

**Comment by rtyler** [comment]
I'm returning to this and I have some exciting developments! Datafusion 44 introduced [LazyMemoryExec](https://docs.rs/datafusion/latest/datafusion/physical_plan/memory/struct.LazyMemoryExec.html) which is the missing piece to make this a _lot_ simpler.

My prior attempts at solving this problem were effectively in a similar solution domain as `LazyMemoryExec`, except I was blurring the lines between Python and Rust more than was probably safe or sane :laughing: 


Conceptually I am building from @aersam's great work in #2289 and using channels to safely bring `RecordBatch`es from the Python layer where there's frankly a _lot_ going on with the Tokio and Python concurrency approaches. With a channel effectively just sending over `RecordBatch` after `RecordBatch`.

On the Rust side, the `LazyMemoryExec` approach makes it more viable to just have a generator receiving these `RecordBatch`es and streaming them into Datafusion.

I hope to have things cleaned up and ready to rock this weekend

**Comment by ion-elgreco** [comment]
Memory has become more stable with: https://github.com/delta-io/delta-rs/pull/3176, https://github.com/delta-io/delta-rs/pull/3141
